### PR TITLE
[Draft] Add Metrics Notification

### PR DESCRIPTION
**Pull request by sowu880**

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Extensions/DictionaryExtensions.cs [hunk 1]
```diff
@@ -30,6 +30,26 @@ public static class DictionaryExtensions
...
         }
 
+        /// <summary>
+        /// Concat two string to long dictionary, for duplicated key, add the their value together.
+        /// For example. the concat result of dic1 {{"key1",1}} and dic2 {{"key1",2}, {"key2", 5}} is {{"key1",3}, {"key2", 5}}
+        /// </summary>
+        /// <param name="dic1">the first dictionary.</param>
+        /// <param name="dic2">the second dictionary.</param>
+        /// <returns>the concat dictionary.</returns>
+        public static Dictionary<string, long> ConcatDictionaryCount(this Dictionary<string, long> dic1, Dictionary<string, long> dic2)
```
Comment (ACMoretxj) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Extensions/DictionaryExtensions.cs:40]: If `dic1` equals to `dic2`, will this method result to infinite loop?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FailedJobNotification.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+using System;
+using System.Collections.Generic;
+using MediatR;
+using Microsoft.Health.Fhir.Synapse.Common.Models.Jobs;
+using Newtonsoft.Json;
+
+namespace Microsoft.Health.Fhir.Synapse.Common.Notification
+{
+    public class FailedJobNotification : INotification
```
Comment (ACMoretxj) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FailedJobNotification.cs:14]: The `FailedJobNotification` and `SuccessfulJobNotification` share many common properties, is it possible to set a common parent notification for them?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Models/Jobs/Job.cs [hunk 3]
```diff
@@ -124,6 +134,12 @@ public class Job
         [JsonProperty("failedReason")]
         public string FailedReason { get; set; }
 
+        /// <summary>
+        /// Failure exception for this job.
+        /// </summary>
+        [JsonProperty("failedException")]
+        public Exception FailedException { get; set; }
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Models/Jobs/Job.cs:141]: Do we have tested this **Exception** can be correctly parsed and serialized between JSON content in blob and runtime **Job** instance?

Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Models/Jobs/Job.cs:141]: From the codes, seems we only want to know the exception type and save that to notifictaion?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FailedJobNotification.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+        /// <summary>
+        /// Started timestamp of job.
+        /// </summary>
+        public DateTimeOffset StartedTime { get; }
+
+        /// <summary>
+        /// Completed timestamp of job.
+        /// </summary>
+        public DateTimeOffset? CompletedTime { get; set; }
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FailedJobNotification.cs:49]: Can this and all fields below be marked only ```get```?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FilterNotification.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        /// <summary>
+        /// The filter information.
+        /// </summary>
+        public FilterInfo FilterInfo { get; }
+
+        /// <summary>
+        /// Processed resource count for each schema type.
+        /// </summary>
+        public Dictionary<string, int> ProcessedResourceCounts { get; set; }
+
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FilterNotification.cs:42]: Additional line

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FilterNotification.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        /// <summary>
+        /// The filter information.
+        /// </summary>
+        public FilterInfo FilterInfo { get; }
+
+        /// <summary>
+        /// Processed resource count for each schema type.
+        /// </summary>
+        public Dictionary<string, int> ProcessedResourceCounts { get; set; }
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Notification/FilterNotification.cs:41]: Can this be marked only ```get```?

File: FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.E2ETests/TestFailedJobNotificationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.Health.Fhir.Synapse.Common.Notification;
+
+namespace Microsoft.Health.Fhir.Synapse.E2ETests
+{
+    public class TestFailedJobNotificationHandler : INotificationHandler<FailedJobNotification>
+    {
+        public Task Handle(FailedJobNotification notification, CancellationToken cancellationToken)
```
Comment (QuanWanxx) [inline FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.E2ETests/TestFailedJobNotificationHandler.cs:15]: Where does this method be used?

Comment (QuanWanxx) [reply FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.E2ETests/TestFailedJobNotificationHandler.cs:15]: Maybe we can set up static variable **E2ETests.IsFailedJobNotificationTriggered** more directly?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/JobManager.cs [hunk 8]
```diff
@@ -77,15 +82,22 @@ public async Task RunAsync(CancellationToken cancellationToken = default)
                 try
                 {
                     job.Status = JobStatus.Running;
+                    job.StartedTime = DateTimeOffset.UtcNow;
                     await _jobExecutor.ExecuteAsync(job, cancellationToken);
                     job.Status = JobStatus.Succeeded;
                     await _jobStore.CompleteJobAsync(job, cancellationToken);
+
+                    await _mediator.Publish(new SuccessfulJobNotification(job), CancellationToken.None);
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/JobManager.cs:90]: Where will this mediator publish the notifications?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/JobManager.cs [hunk 9]
```diff
@@ -77,15 +82,22 @@ public async Task RunAsync(CancellationToken cancellationToken = default)
...
+                    await _mediator.Publish(new FilterNotification(job), CancellationToken.None);
                 }
                 catch (Exception exception)
                 {
                     job.Status = JobStatus.Failed;
                     job.FailedReason = exception.ToString();
+                    job.FailedException = exception;
+
                     await _jobStore.CompleteJobAsync(job, cancellationToken);
+                    await _mediator.Publish(new FailedJobNotification(job), CancellationToken.None);
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/JobManager.cs:100]: Maybe we can add unit tests for publishing successful or failed notification when processing the job.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Tasks/TaskExecutor.cs [hunk 10]
```diff
@@ -488,6 +491,8 @@ private void AddFhirResourcesToCache(List<JObject> fhirResources, CacheResult ca
                                 skippedCount);
                         }
 
+                        taskContext.OutputCount =
+                            taskContext.OutputCount.AddToDictionary(resourceType, parquetStream.BatchSize);
```
Comment (QuanWanxx) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Tasks/TaskExecutor.cs:495]: Nit: Can you help to rename the **BatchSize** in **StreamBatchData** to another like **BatchCount**? Now we have the output size, feel that name may leading to misunderstand.

**Comment by moria97** [comment]
@sowu880 Is this pr still active?

**Comment by sowu880** [comment]
Abandon notification component temporary.

**Pull request closed without merging**